controllers/clustertopology_controller_compute.go

```diff
@@ -140,6 +141,149 @@ func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clust
 	return class, nil
 }
 
+// getCurrentState gets information about a cluster topology by inspecting the state of the cluster infrastructure,
+// the control plane, and the machine deployments associated with the cluster.
+func (r *ClusterTopologyReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster, class clusterv1.ClusterClass) (*clusterTopologyState, error) {
+	var err error
+	var infra *unstructured.Unstructured
+	var cp controlPlaneTopologyState
+	var m []machineDeploymentTopologyState
+	// Reference to the cluster infrastructure can be nil and is expected to be on the first reconcile loop
+	// This reference can be nil and the method should still be allowed to continue
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err = r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, errors.Wrap(err, "current cluster state not found")
+		}
+	}
+	// Reference to the control plane can be nil, and if the reference to the control plane is not nil the reference
+	// to the control plane infrastructure may or may not exist. If the linked ClusterClass requires control plane
+	// infrastructure this will throw an unrecoverable error due to the configuration mismatch.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err = r.getControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, errors.Wrap(err, "current cluster state not found")
+		}
+	}
+	// This reference can be empty and should not return an error if there are not Machine Deployments associated with
+	// the cluster at time of requesting state. If the machine deployments are missing information about their own
+	// infrastructure and bootstrap references this will throw an error.
+	m, err = r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "current cluster state not found")
+	}
+	return &clusterTopologyState{
+		cluster:               cluster,
+		infrastructureCluster: infra,
+		machineDeployments:    m,
+		controlPlane:          cp,
+	}, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the infrastructure underlying the cluster. If infrastructure is nil
+// no error is thrown and reconciliation continues. If a reference is set but not found, through an error or not existing
+// in the api server an error is thrown.
```

Comment:
nit
```suggestion
// getCurrentInfrastructureClusterState looks for the infrastructure underlying the cluster; If a reference is not found, 
// an error is thrown.
```
This is technically correct, but given that infrastructure ref nil is checked outside of this func, I think we should revmove this part from the comment